graph/src/data/store/scalar.rs

```diff
@@ -10,14 +13,158 @@ use stable_hash::{
 };
 use std::convert::{TryFrom, TryInto};
 use std::fmt::{self, Display, Formatter};
+use std::io::Write;
 use std::ops::{Add, Div, Mul, Rem, Sub};
 use std::str::FromStr;
 
 pub use num_bigint::Sign as BigIntSign;
 
-// Caveat: The exponent is currently an i64 and may overflow.
-// See https://github.com/akubera/bigdecimal-rs/issues/54.
-pub type BigDecimal = bigdecimal::BigDecimal;
+/// All operations on `BigDecimal` return a normalized value.
+// Caveat: The exponent is currently an i64 and may overflow. See
+// https://github.com/akubera/bigdecimal-rs/issues/54.
+// Using `#[serde(from = "BigDecimal"]` makes sure deserialization calls `BigDecimal::new()`.
+#[derive(
+    Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, AsExpression, FromSqlRow,
+)]
+#[serde(from = "bigdecimal::BigDecimal")]
+#[sql_type = "diesel::sql_types::Numeric"]
+pub struct BigDecimal(bigdecimal::BigDecimal);
+
+impl From<bigdecimal::BigDecimal> for BigDecimal {
+    fn from(big_decimal: bigdecimal::BigDecimal) -> Self {
+        BigDecimal(big_decimal).normalize()
+    }
+}
+
+impl BigDecimal {
+    pub fn new(digits: BigInt, exp: i64) -> Self {
+        Self::from(bigdecimal::BigDecimal::new(digits.0, -exp))
+    }
+
+    pub fn as_bigint_and_exponent(&self) -> (num_bigint::BigInt, i64) {
+        self.0.as_bigint_and_exponent()
+    }
+
+    pub(crate) fn digits(&self) -> u64 {
+        self.0.digits()
+    }
+
+    // Copy-pasted from `bigdecimal::BigDecimal::normalize`. We can use the upstream version once it
+    // is included in a released version supported by Diesel.
+    pub fn normalize(&self) -> BigDecimal {
+        let (bigint, exp) = self.0.as_bigint_and_exponent();
+        let (_, mut digits) = bigint.to_u32_digits();
+        let mut digits_iter = digits.iter().rev();
+        let mut trailing_count = 0;
+        while digits_iter.next() == Some(&0) {
+            trailing_count += 1;
+        }
+        digits.truncate(digits.len() - trailing_count as usize);
```

Comment:
I copied the tests from the [bigdecimal PR](https://github.com/akubera/bigdecimal-rs/commit/03f56367e27b3f4fcaf70b117bcd6fc49c64593a#diff-b4aea3e418ccdb71239b96952d9cddb6R2822) into this repo, and they failed. So, something is awry.

Here's what I got...
```
thread 'data::store::scalar::test::test_normalize' panicked at 'assertion failed: `(left == right)`
  left: `BigDecimal(BigDecimal("1000"))`,
 right: `BigDecimal(BigDecimal("10"))`', graph/src/data/store/scalar.rs:594:13
```

This works if the exponent is not negated in our newtype constructor.

Reply:
I copied over the test for good measure, but they were testing fine (though they were wrong as you noticed below). Because our constructor handles the exponent in the more natural way while bigdecimal flips it, the test needs to be adapted for our constructor.